prereise/gather/griddata/hifld/data_process/transmission.py

```diff
@@ -27,6 +29,162 @@ def check_for_location_conflicts(substations):
+    n_unique_orphan_endpoint = len(unique_endpoint_coord - set(subcoord2subid))
+    n_unconnected_sub = len(set(subcoord2subid) - unique_endpoint_coord)
+
+    print(
+        f"there are {n_two_sub_missing} lines with two substations missing\n"
+        f"there are {n_one_sub_missing} lines with one substation missing\n"
+        f"there are {n_unique_orphan_endpoint} unique orphan endpoint\n"
+        f"there are {n_unconnected_sub} substations unconnected\n"
+    )
+
+    return line2sub
+
+
+def assign_substations_to_lines(substations, lines, line2sub):
+    """Locate closest substation of unconnected lines and update information
+    information of lines. Note that the ``line2sub`` data frame is modified in-place.
+
+    :param pandas.DataFrame substations: data frame of substations.
+    :param pandas.DataFrame lines: data frame of all lines.
+    :param pandas.DataFrame line2sub: indices: table mapping lines to substations as
+        returned by :func:`map_lines_to_substations_using_coords`.
+    :return: (*pandas.DataFrame*) -- original lines data frame with updated *SUB_1*
+        and *SUB_2* columns, new *SUB_1_ID*, *SUB_2_ID* along with *OTHER_SUB_1_ID*
+        and *OTHER_SUB_2_ID* when multiple location are associated to the same endpoint.
+    """
+    print("finding closest neighbor to unconnected lines")
+
+    # find closest neighbor of unconnected lines
+    subcoord2subid = substations.groupby(["LONGITUDE", "LATITUDE"]).groups
+    subcoord = list(subcoord2subid)
+    missing = line2sub.query("FROM_SUB.isnull() or TO_SUB.isnull()").index
+    endpoint2neighbor = {}
+    for i in tqdm(missing, total=len(missing)):
+        for endpoint in ["FROM", "TO"]:
+            if isinstance(line2sub.loc[i, f"{endpoint}_SUB"], float):
+                if line2sub.loc[i, endpoint] in endpoint2neighbor:
+                    line2sub.loc[i, f"{endpoint}_SUB"] = endpoint2neighbor[
+                        line2sub.loc[i, endpoint]
+                    ]
+                else:
+                    subid = subcoord2subid[
+                        subcoord[
+                            find_closest_neighbor(line2sub.loc[i, endpoint], subcoord)
+                        ]
+                    ]
+                    line2sub.loc[i, f"{endpoint}_SUB"] = list(subid)
+                    endpoint2neighbor[line2sub.loc[i, endpoint]] = list(subid)
+
+    # select all substations
+    new_substations = substations.copy()
+    new_substations = substations.loc[
+        set(
+            itertools.chain(
+                *{tuple(i) for j in ["FROM_SUB", "TO_SUB"] for i in line2sub[j]}
+            )
+        )
+    ]
+
+    # pick unique substation at each end. Order of preference:
+    # SUBSTATION --> TAP --> RISER --> DEAD END
+    all2one = {
+        tuple(i): None
+        for j in ["FROM_SUB", "TO_SUB"]
+        for i in line2sub[j]
+        if len(i) > 1
+    }
+    for all in all2one:
```

Comment:
Should we rename this to avoid overwriting the python built-in `all()`? It may not make a difference right now, but could interfere with this uses of `all()` within this functions namespace in a potential future refactor.

Reply:
good point

Reply:
Done